zkml/src/layers/transformer/mha.rs

```diff
@@ -54,52 +57,198 @@ impl OpInfo for MhaQK {
+        &self,
+        input_shapes: &[Vec<usize>],
+        padding_mode: PaddingMode,
+    ) -> Vec<Vec<usize>> {
+        let seq_len = input_shapes[1][0];
+        let v_shape = match padding_mode {
+            PaddingMode::NoPadding => {
+                vec![self.num_heads, seq_len, self.head_dim]
             }
             PaddingMode::Padding => {
                 vec![
-                    vec![
-                        self.num_heads,
-                        q_len.next_power_of_two(),
-                        seq_len.next_power_of_two(),
-                    ],
-                    vec![
-                        self.num_heads,
-                        seq_len.next_power_of_two(),
-                        self.head_dim.next_power_of_two(),
-                    ],
+                    self.num_heads.next_power_of_two(),
+                    seq_len.next_power_of_two(),
+                    self.head_dim.next_power_of_two(),
                 ]
             }
-        }
+        };
+        assert_eq!(
+            input_shapes[0][2], seq_len,
+            "qk should have the same sequence length as v"
+        );
+        let mul_input_shapes = vec![
+            input_shapes[0].clone(), // QK
+            v_shape,                 // V
+        ];
+        self.mul.output_shapes(&mul_input_shapes, padding_mode)
+    }
+
+    fn describe(&self) -> String {
+        self.mul.describe()
+    }
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct Mha<N> {
+    linear: MhaQk,
+    softmax: Softmax<N>,
+    final_mul: MhaFinalMul,
+    final_reshape: Reshape,
+}
+
+impl<N: Number> Mha<N> {
+    pub fn new(num_heads: usize, head_dim: usize) -> anyhow::Result<Self> {
+        let linear = MhaQk {
+            num_heads,
+            head_dim,
+        };
+        let softmax = Softmax::new()
+            .with_scale(N::from_f32((1.0 / (head_dim as f32)).sqrt())?)
+            .on_dim(1);
+        let final_mul = MhaFinalMul::new(num_heads, head_dim);
+        // reshape the output from [q_len, num_heads, head_dim] to [q_len, num_heads*head_dim]
+        let final_reshape = Reshape::new_subspace(1..=2, vec![num_heads * head_dim]);
+        Ok(Self {
+            linear,
+            softmax,
+            final_mul,
+            final_reshape,
+        })
+    }
+
+    pub(crate) fn evaluate_with_softmax_out<E: ExtensionField>(
+        &self,
+        inputs: &[&Tensor<N>],
+        unpadded_input_shapes: Vec<Vec<usize>>,
+    ) -> anyhow::Result<(LayerOut<N, E>, LayerOut<N, E>)>
+    where
+        Softmax<N>: Evaluate<N>,
+    {
+        let unpadded_input_shapes = if unpadded_input_shapes.is_empty() {
+            // take input shapes from inputs
+            inputs.iter().map(|input| input.get_shape()).collect()
+        } else {
+            unpadded_input_shapes
+        };
+
+        let linear_out = self
+            .linear
+            .evaluate::<E>(&inputs[..2], unpadded_input_shapes[..2].to_vec())?;
```

Comment:
a little `ensure!(inputs.len() >= 2)` would be nice there, wdyt ?

Reply:
Done in commit https://github.com/Lagrange-Labs/deep-prove/pull/156/commits/1d0caf95c67822fc7bdb0a44fb51609c9f68e538